DirectProgramming/C++SYCL_FPGA/ReferenceDesigns/matmul/README.md

```diff
@@ -0,0 +1,323 @@
+
+| Optimized for        | Description
+|:---                  |:---
+| OS                   | Ubuntu* 18.04/20.04 <br> RHEL*/CentOS* 8 <br> SUSE* 15 <br> Windows* 10
+| Hardware             | Intel® Agilex™, Arria® 10, and Stratix® 10 FPGAs
+| Software             | Intel® oneAPI DPC++/C++ Compiler
+
+> **Note**: Even though the Intel DPC++/C++ OneAPI compiler is enough to compile for emulation, generating reports and generating RTL, there are extra software requirements for the simulation flow and FPGA compiles.
+>
+> For using the simulator flow, Intel® Quartus® Prime Pro Edition and one of the following simulators must be installed and accessible through your PATH:
+> - Questa*-Intel® FPGA Edition
+> - Questa*-Intel® FPGA Starter Edition
+> - ModelSim® SE
+>
+> When using the hardware compile flow, Intel® Quartus® Prime Pro Edition must be installed and accessible through your PATH.
+>
+> :warning: Make sure you add the device files associated with the FPGA that you are targeting to your Intel® Quartus® Prime installation.
+
+### Performance
+
+Performance results are based on testing as of March 6, 2023.
+
+> **Note**: Refer to the [Performance Disclaimers](/DirectProgramming/C++SYCL_FPGA/README.md#performance-disclaimers) section for important performance information.
+
+| Device                                            | Throughput
+|:---                                               |:---
+| Intel® PAC with Intel® Arria® 10 GX FPGA          | 75k matrices/s for matrices of size 64 * 64
+| Intel® FPGA PAC D5005 (with Intel Stratix® 10 SX) | 90k matrices/s for matrices of size 64 * 64
+
+
+## Key Implementation Details
+
+The matrix multiplication algorithm multiplies two single-precision floating-point matrices *A* (*m × n*) and *B* (*n × p*) to produce matrix *C* (*m × p*). The algorithm is implemented using a systolic array approach, employing an array of *m<sub>PE</sub> × p<sub>PE</sub>* processing elements (PEs), each of which is responsible for computing a dot product corresponding to an element of the output matrix. As a result, this implementation computes the matrix multiplication in tiles of *m<sub>PE</sub> × p<sub>PE</sub>* at a time.
+
+<p align="center">
+  <img src=assets/overview.png />
+</p>
+
+Each iteration, a PE consumes inputs from previous PEs in the same row or column and uses these data to compute a multiply-accumulate. It then passes the inputs along to the next PE. We orchestrate the flow of data through the systolic array such that each PE is responsible for computing one element of the output matrix. Using this approach, it takes a total of *n* iterations to compute one matrix product, where every iteration, we compute a partial sum for each of the elements of the output matrix.
```

Comment:
Right, after _n_ cycles the full output matrix will be computed and stored in the compute kernel's registers, but it takes a further _p_ cycles to drain the matrix out. So it will actually be _n+p_ cycles between when the first set of values is read in to when the last set of values is written out for any given matrix. However since we can begin computing the next matrix while the previous one drains out, the extra iterations would be hidden by the next repetition (except the last one). I'll make this more clear in the Readme